pkg/ddl/index_merge_tmp.go

```diff
@@ -172,71 +174,101 @@ func (w *mergeIndexWorker) setCurrentIndexForRange(taskRange *reorgBackfillTask)
+			txn.SetOption(kv.ResourceGroupName, w.jobContext.resourceGroupName)
 
-		for i, idxRecord := range tmpIdxRecords {
-			taskCtx.scanCount++
-			// The index is already exists, we skip it, no needs to backfill it.
-			// The following update, delete, insert on these rows, TiDB can handle it correctly.
-			// If all batch are skipped, update first index key to make txn commit to release lock.
-			if idxRecord.skip {
-				continue
+			tmpIdxRecords, nextKey, taskDone, err := w.fetchTempIndexVals(txn, &taskCtx.scanCount, taskRange)
+			if err != nil {
+				return errors.Trace(err)
 			}
+			taskCtx.nextKey = nextKey
+			taskCtx.done = taskDone
 
-			// Lock the corresponding row keys so that it doesn't modify the index KVs
-			// that are changing by a pessimistic transaction.
-			rowKey := tablecodec.EncodeRecordKey(w.table.RecordPrefix(), idxRecord.handle)
-			err := txn.LockKeys(context.Background(), new(kv.LockCtx), rowKey)
+			err = w.batchCheckTemporaryUniqueKey(txn, tmpIdxRecords)
 			if err != nil {
 				return errors.Trace(err)
 			}
 
-			if idxRecord.delete {
-				if idxRecord.unique {
-					err = txn.GetMemBuffer().DeleteWithFlags(w.originIdxKeys[i], kv.SetNeedLocked)
+			for i, idxRecord := range tmpIdxRecords {
+				// The index is already exists, we skip it, no needs to backfill it.
+				// The following update, delete, insert on these rows, TiDB can handle it correctly.
+				// If all batch are skipped, update first index key to make txn commit to release lock.
+				if idxRecord.skip {
+					continue
+				}
+
+				originIdxKey := w.originIdxKeys[i]
+				if idxRecord.delete {
+					err = txn.GetMemBuffer().Delete(originIdxKey)
 				} else {
-					err = txn.GetMemBuffer().Delete(w.originIdxKeys[i])
+					err = txn.GetMemBuffer().Set(originIdxKey, idxRecord.vals)
+				}
+				if err != nil {
+					return err
 				}
-			} else {
-				err = txn.GetMemBuffer().Set(w.originIdxKeys[i], idxRecord.vals)
+
+				err = txn.GetMemBuffer().Delete(w.tmpIdxKeys[i])
+				if err != nil {
+					return err
+				}
+
+				failpoint.InjectCall("mockDMLExecutionMergingInTxn")
+
+				taskCtx.addedCount++
 			}
-			if err != nil {
-				return err
+			return nil
+		})
+		if err != nil {
+			if kv.IsTxnRetryableError(err) {
+				if err := w.ddlCtx.isReorgRunnable(ctx, false); err != nil {
+					return taskCtx, errors.Trace(err)
+				}
+				if w.batchCnt > 1 {
+					w.batchCnt /= 2
+				}
+				w.conflictCounter.Add(1)
+				backoff := kv.BackOff(uint(attempts))
+				logutil.DDLLogger().Warn("temp index merge worker retry",
+					zap.Int64("jobID", taskRange.jobID),
+					zap.Int("batchCnt", w.batchCnt),
+					zap.Int("attempts", attempts),
+					zap.Duration("backoff", time.Duration(backoff)),
+					zap.Uint64("startTS", currentTxnStartTS),
+					zap.Error(err))
+				continue
 			}
-			taskCtx.addedCount++
+			return taskCtx, errors.Trace(err)
 		}
-		return nil
-	})
+		if attempts <= 1 {
+			w.batchCnt = min(int(vardef.GetDDLReorgBatchSize()), w.batchCnt*2)
```

Comment:
If attempts <= 1, it follows the original logic. So, can this assignment be moved before the for loop, and could the condition check be removed?"​